File: Algo.Analytics.FSharp/PriceVolumeScript.fs
namespace StockSharp.Algo.Analytics

open System
open System.Linq
open System.Threading
open System.Threading.Tasks
open System.Collections.Generic

open Ecng.Common
open Ecng.Drawing
open Ecng.Logging

open FSharp.Control

open StockSharp.Algo.Analytics
open StockSharp.Algo.Storages
open StockSharp.Algo.Candles
open StockSharp.Messages

/// <summary>
/// The analytic script, calculating distribution of the volume by price levels.
/// </summary>
type PriceVolumeScript() =
    interface IAnalyticsScript with
        member this.Run
            (
                logs: ILogReceiver,
                panel: IAnalyticsPanel,
                securities: SecurityId[],
                fromDate: DateTime,
                toDate: DateTime,
                storage: IStorageRegistry,
                drive: IMarketDataDrive,
                format: StorageFormats,
                dataType: DataType,
                cancellationToken: CancellationToken
            ) : Task =

            task {
                if securities.Length = 0 then
                    logs.LogWarning("No instruments.")
                else
                    // This script processes only the first instrument
                    let security = securities.First()

                    logs.LogInfo("Processing {0}...", security)

                    // Get candle storage
                    let candleStorage = storage.GetCandleMessageStorage(security, dataType, drive, format)

                    // Get available dates within the specified period
                    let! datesSeq = candleStorage.GetDatesAsync(fromDate, toDate, cancellationToken)
                    let dates = datesSeq.ToArray()

                    if dates.Length = 0 then
                        logs.LogWarning("no data")
                    else
                        // Group candles by "middle price" = LowPrice + (HighPrice - LowPrice) / 2
                        let rows = Dictionary<decimal, decimal>()
                        let mutable prevDate = DateOnly.MinValue

                        do! candleStorage.LoadAsync(fromDate, toDate)
                            |> TaskSeq.iter (fun candle ->
                                let currDate = DateOnly.FromDateTime(candle.OpenTime.Date)
                                if currDate <> prevDate then
                                    prevDate <- currDate
                                    logs.LogInfo("  {0}...", currDate)

                                let midPrice = candle.LowPrice + (candle.GetLength() / 2m)
                                match rows.TryGetValue(midPrice) with
                                | true, volume -> rows.[midPrice] <- volume + candle.TotalVolume
                                | false, _ -> rows.[midPrice] <- candle.TotalVolume
                            )

                        // Draw histogram on the chart
                        panel.CreateChart<decimal, decimal>()
                             .Append(security.ToStringId(), rows.Keys, rows.Values, DrawStyles.Histogram)
            }

File: Algo.Analytics.CSharp/PriceVolumeScript.cs
namespace StockSharp.Algo.Analytics;

/// <summary>
/// The analytic script, calculating distribution of the volume by price levels.
/// </summary>
public class PriceVolumeScript : IAnalyticsScript
{
	async Task IAnalyticsScript.Run(ILogReceiver logs, IAnalyticsPanel panel, SecurityId[] securities, DateTime from, DateTime to, IStorageRegistry storage, IMarketDataDrive drive, StorageFormats format, DataType dataType, CancellationToken cancellationToken)
	{
		if (securities.Length == 0)
		{
			logs.LogWarning("No instruments.");
			return;
		}

		// script can process only 1 instrument
		var security = securities.First();

		logs.LogInfo("Processing {0}...", security);

		// get candle storage
		var candleStorage = storage.GetCandleMessageStorage(security, dataType, drive, format);

		// get available dates for the specified period
		var dates = await candleStorage.GetDatesAsync(from, to).ToArrayAsync(cancellationToken);

		if (dates.Length == 0)
		{
			logs.LogWarning("no data");
			return;
		}

		// grouping candles by middle price
		var rows = new Dictionary<decimal, decimal>();
		var prevDate = default(DateOnly);

		await foreach (var candle in candleStorage.LoadAsync(from, to).WithCancellation(cancellationToken))
		{
			var currDate = DateOnly.FromDateTime(candle.OpenTime.Date);
			if (currDate != prevDate)
			{
				prevDate = currDate;
				logs.LogInfo("  {0}...", currDate);
			}

			var midPrice = candle.LowPrice + candle.GetLength() / 2;

			if (rows.TryGetValue(midPrice, out var volume))
				rows[midPrice] = volume + candle.TotalVolume;
			else
				rows[midPrice] = candle.TotalVolume;
		}

		// draw on chart
		panel.CreateChart<decimal, decimal>()
			.Append(security.ToStringId(), rows.Keys, rows.Values, DrawStyles.Histogram);
	}
}